File: java/java-samples/src/main/java/tech/techstay/str/RegularExpression.java
package tech.techstay.str;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class RegularExpression {
  public static void main(String[] args) {
    // the pattern that matches all words starting with b and ending with c
    Pattern pattern = Pattern.compile("\\bb\\w*c\\b");
    String text = "abc bbc cbc bbcc bcd";

    Matcher result = pattern.matcher(text);

    while (result.find()) {
      System.out.println(result.group());
    }
  }
}

File: scala/scala-example/src/main/scala/tech/techstay/std/RegularExpression.scala
package tech.techstay.std

import scala.util.matching.Regex

@main def regularExpression(): Unit =
  val pattern: Regex = """\bb+c\b""".r
  val text = "abc bbc cbc bbbbbbc"
  for i <- pattern.findAllMatchIn(text) do println(i.group(0))

  // regex pattern matching
  def filterName(text: String): Unit =
    val namePattern = """(\w+)""".r
    val fullNamePattern = """(\w+)[\s,|:]*(\w+)""".r
    text match
      case namePattern(name) => println(s"Your name: $name")
      case fullNamePattern(firstname, lastname) =>
        println(s"Your full name: $firstname $lastname")
      case _ => println("Invalid format")

  filterName("techstay")
  filterName("Mike Clinton")
  filterName("Foo, Bar")
